### ISSUE TITLE
Improved interface for form fields

### ISSUE THREAD

**Issue by mever** [qooxdoo/qooxdoo#9366]
Dear developers,

The form components, great as they are, leave some room for improvement (as do all great engineering efforts). Therefore I want to pitch an idea for improving the form package. Goal is to leave existing code backwards compatible and add functionality that make dealing with the form easier.
Conceptually we have two categories of classes, the form and its fields. Functions of the form; like resetting, validation and data binding now rely on the specifics of those fields. When I browse the code I see a lot of if statements for specific uses that make the code fragile. I want to exercise the dependency inversion principle and the interface segregation principle from SOLID to make this better.

I think a new interface in the form package called IField with the following methods should do the trick:
```
+ setInputValue(value anytype|null) void
+ resetInputValue() void
+ getInputValue() anytype|promise|null

// setInputValue updates the view and model value
// resetInputValue updates the view with model value
// getInputValue updates the model with view value and return the model value

// The view value is what is displayed, the model value is what's in the memory of the field.

// getInputValue may return a promise in order to allow fields to be asynchronous (e.g. fetch server data)
```
_(input value is chosen because value is taken and BC is a priority)_

With this new interface resetting and data binding is possible in a field generic way. Validation may require an additional interface (e.g. IFieldValidation) because each field may choose to present feedback about the (invalid) value specifically. My current position is to leave it compatible with IForm.

This proposal makes it easier to create more fields that can be used in the form.

Any thoughts?

**Comment by derrell** [comment]
This looks really good in principle! Have you done code changes to see that it really can be done in a backwards-compatible fashion? I would be very eager to see a PR that implements this.

**Comment by level420** [comment]
@mever could you please give a practical example where your proposal helps? Thank you!

BTW.: it took some time until I found out that BC does not mean "before christ" ;-)

**Comment by mever** [comment]
@level420 I created a separate form manager which implements a resetFields method. This method calls the form.reset() method on the Qooxdoo component and some custom logic to make it work with our custom fields. I like to get rid of resetFields and let Qooxdoo handle it.

**Comment by johnspackman** [comment]
So is this a unification of the `qx.ui.form.IForm`/`IStringForm`/`INumberForm` fields?  IE `qx.ui.form.IForm` sounds like `IFieldValidation`, and `qx.ui.form.IStringForm`/etc would inherit from `qx.ui.form.IField`?

**Comment by mever** [comment]
@johnspackman yes.

This interface allows complex fields to be accessed. The value of a selectbox is the selected item. The value of a multi-selection list or group of checkboxes is an array of model values. Even nested forms are possible as field when they implement the interface.

**Comment by level420** [comment]
Ah OK. I've done something similar (without defining an interface for that). For example I've implemented a form field `qx.ui.form.CheckButtonBox` which is a layout widget holding a bunch of check boxes and implementing getValue, setValue etc. where the value is an array representing the checked state of the buttons in that check button box.

**Comment by mever** [comment]
It is hard to understand the form component and add new fields. Currently we have the most problems with the resetter component because it does not relies on a single interface but instead relies on implementation details (i.e. the resetter should not have to know that a field works with a selection).
Earlier I pushed for IArrayForm and IModelForm (https://github.com/qooxdoo/qooxdoo/pull/9075), which we're using successfully and in production. But after studying the SOLID principles I came to the conclusion that a single interface is even better.

The project I am working on is very big and contains a lot of code (around 400 Qooxdoo classes). And I do not want it to become a fork of Qooxdoo. That's why I try to push improvements and new components upstream instead of making competing components.

**Comment by mever** [comment]
Can we allow setValue to cause a minor BC break? The API doc marks the return value of AbstractField.setValue as void. But when I look at the implementation it returns particular values.

**Comment by level420** [comment]
@mever what would the BC (I learned!) be? To change it to not return a value or to add the correct jsdoc part to document it does return a value?

**Comment by mever** [comment]
@level420 the interface I've build returns null or an Error object for setValue. Clients that expect undefined (void) or the behavior implemented in [AbstractField.setValue](https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/ui/form/AbstractField.js#L635) (when a string is returned) would break.
Although I do not think a lot of clients rely on the return value of setValue I do not want to miss something.

**Comment by level420** [comment]
closed via PR #9367
